### DISCUSSION TITLE
How to get mousedown position in ui.matplotlib

### DISCUSSION THREAD

**Discussion by nghia-vo** [zauberzeug/nicegui#3063] [Q&A] [answered]
### Question

I would like to get the mouse position when clicking on an image displayed by ui.matplotlib. How can I achieve that?

```python
import numpy as np
from nicegui import ui, events

def mouse_handler(e: events.MouseEventArguments):
    # How to get mouse position here
    ui.notify("Mouse position is")

main_plot = ui.matplotlib(figsize=(8, 6))
main_plot.on("click", mouse_handler)
mat = np.random.rand(100, 100)
ax = main_plot.figure.gca()
ax.imshow(mat)
main_plot.update()
ui.run()
```

**Comment by falkoschindler** [comment, marked as answer]
Hi @nghia-vo,

The event arguments `e` contain various attributes:
```py
with ui.matplotlib(figsize=(8, 6)).on('click', lambda e: print(e.args)).figure as fig:
    fig.gca().imshow(np.random.rand(100, 100))
```
Output:
```
{'isTrusted': True, '_vts': 1715579733809, 'pointerId': 1, 'width': 1, 'height': 1, 'pressure': 0, 'tiltX': 0, 'tiltY': 0, 'azimuthAngle': 0, 'altitudeAngle': 1.5707963267948966, 'tangentialPressure': 0, 'twist': 0, 'pointerType': 'mouse', 'isPrimary': False, 'screenX': 464, 'screenY': 314, 'clientX': 464, 'clientY': 189, 'ctrlKey': False, 'shiftKey': False, 'altKey': False, 'metaKey': False, 'button': 0, 'buttons': 0, 'relatedTarget': None, 'pageX': 464, 'pageY': 189, 'x': 464, 'y': 189, 'offsetX': 449, 'offsetY': 173, 'movementX': 0, 'movementY': 0, 'fromElement': None, 'toElement': None, 'layerX': 464, 'layerY': 189, 'detail': 1, 'sourceCapabilities': {}, 'which': 1, 'type': 'click', 'eventPhase': 3, 'bubbles': True, 'cancelable': True, 'defaultPrevented': False, 'composed': True, 'timeStamp': 583.8999999761581, 'returnValue': True, 'cancelBubble': False, 'NONE': 0, 'CAPTURING_PHASE': 1, 'AT_TARGET': 2, 'BUBBLING_PHASE': 3}
```

Unfortunately the mouse coordinates are in screen pixels and not in the data coordinate system 0..100. I'm not sure if there's a way to convert them reliably though.

**Reply by nghia-vo** [reply]
Thanks Falko, that's useful information. I can go from here.

**Reply by moi90** [reply]
@nghia-vo Were you able to convert the screen coordinates to data coordinates?

**Reply by nghia-vo** [reply]
Yes, here is the link to the implementation: https://github.com/zauberzeug/nicegui/discussions/3380

**Reply by nghia-vo** [reply]
https://github.com/algotom/broh5/blob/f7dfe6e72859e40c2ec529eb1dd5ab0d62641c12/broh5/lib/interactions.py#L121

**Reply by evnchn** [reply]
👀 may see how to integrate with NiceGUI. In my ideas list now

**Reply by evnchn** [reply]
Reviewing, the line for figuring out the data is just:  

https://github.com/algotom/broh5/blob/f7dfe6e72859e40c2ec529eb1dd5ab0d62641c12/broh5%2Flib%2Finteractions.py#L114-L119

Then, it isn't as complicated as I thought (I thought we had to do some complex math and/or hacky things to figure it out)

If that's the case, this better go into the Wiki than into NiceGUI itself. 

Also, would be kind of unfair that only matplotlib got this functionality if I'm putting it in NiceGUI.

**Reply by falkoschindler** [reply]
> If that's the case, this better go into the Wiki than into NiceGUI itself.

Or maybe a new demo at https://nicegui.io/documentation/matplotlib?

**Reply by nghia-vo** [reply]
It's not that simple:
```python
def __get_xy(self, x, y, ax):
    try:
        xn, yn = ax.transData.inverted().transform((x, y))
        return xn, yn
    except Exception as e:
        return None, None
x, y = self.__get_xy(e.args['offsetX'], e.args['offsetY'], self.ax)
(height, width) = self.image.shape
x = int(x)
y_max = self.ax.transData.inverted().transform(
    (0, self.ax.get_ylim()[-1]))[-1]
y = height - 1 - int(y) + int((y_max - height) / 2
```
The y-position is a tricky one to get it right. It took me a while to figure out.

**Reply by moi90** [reply]
> Also, would be kind of unfair that only matplotlib got this functionality if I'm putting it in NiceGUI.

What do you mean? It would be unfair if only `ui.matplotlib` would be able to get the mouse position and other plotting widgets (`ui.echart`, `ui.plotly`) don't? Unfair for whom?

I think this would be really useful and it would be great to have a `matplotlib:click` event (or something similar) that directly provides the position in [figure, axes, data]-coordinates.

**Reply by evnchn** [reply]
In retrospect, it's fine if `ui.matplotlib` gets the data position first, not really "unfair" to other plotting widgets (ui.echart, ui.plotly)

But, since the technique is already laid-out, not sure of the demand to put it into NiceGUI. Also would probably be a nightmare to test compatibility with all charts.

**Reply by falkoschindler** [reply]
> Also would probably be a nightmare to test compatibility with all charts.

You mean plot, scatter, image, etc.? I agree. The snippet from https://github.com/zauberzeug/nicegui/discussions/3063#discussioncomment-12830458 seems to require an image, which is not very generic. How would we get data coordinates for a simple `ax.plot(...)`?

An MRE might help to estimate the complexity and to see how it could be integrated.

**Comment by moi90** [comment]
When trying to solve a similar problem, I used `ui.interactive_image` instead and displayed individual components of the figure as SVG shapes.

(This does not provide everything that `matplotlib` provides, but could suffice in simple cases. And then there is the possibility to save a figure as svg and display that. https://matplotlib.org/3.1.1/gallery/user_interfaces/svg_histogram_sgskip.html)
